src/libraries/System.Private.CoreLib/src/System/Type.cs

```diff
@@ -438,11 +438,59 @@ public static Type[] GetTypeArray(object[] args)
             return cls;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static TypeCode GetTypeCode(Type? type)
         {
+            if (RuntimeHelpers.IsKnownConstant(type) && type is RuntimeType)
+            {
+                return GetRuntimeTypeCode((RuntimeType)type);
+            }
             return type?.GetTypeCodeImpl() ?? TypeCode.Empty;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static TypeCode GetRuntimeTypeCode(RuntimeType type)
+        {
+            RuntimeType underlyingType = type;
+            if (type.IsActualEnum)
+                underlyingType = (RuntimeType)type.GetEnumUnderlyingType();
+
+            if (underlyingType == typeof(sbyte))
```

Comment:
This looks slower than the old version which compared integers using a switch statement, this compares
type objects, also on AOT platforms typeof() is more expensive since it needs to load
the object from somewhere.

Reply:
The value is cached after the first call with a type, also on CoreCLR and NativeAOT a type check like this will just load 2 pointers and compare them.

Reply:
@MichalPetryka it won't compare 2 pointers on NativeAOT (and crossgen2), you're referring to the jit specific optimization